### PR TITLE
Initial MPT implementation (2.x)

### PR DESCRIPTION
**Pull request by fyrchik**
This is initial MPT implementation, containing most of the necessary methods.
Storage key transformation to NEO format will be done during integration.
It should probably be left in this branch, not merged.

### CODE REVIEW

**Comment by roman-khimov** [comment]
We won't merge it to master-2.x as of yet, this branch should probably be renamed and `neox-2.x` made protected one.

File: pkg/core/mpt/branch.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+const (
+	// ChildrenCount represents a number of children of a branch node.
+	ChildrenCount = 17
+	// LastChild is the index of the last child.
+	LastChild = ChildrenCount - 1
+)
+
+// BranchNode represents MPT's branch node.
+type BranchNode struct {
+	hash *util.Uint256
```
Comment (roman-khimov) [inline pkg/core/mpt/branch.go:17]: Do we need to use a pointer here?

File: pkg/core/mpt/node.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+const (
+	BranchT    NodeType = 0x00
+	ExtensionT NodeType = 0x01
+	HashT      NodeType = 0x02
+	LeafT      NodeType = 0x03
+)
+
+// NodeWithType represents Node together with it's type.
+type NodeWithType struct {
+	typ NodeType
```
Comment (roman-khimov) [inline pkg/core/mpt/node.go:24]: I don't think you need this field.

File: pkg/core/mpt/branch.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/mpt/branch.go:17]: And in other structures as well. I'd like to minimize microallocations where possible, these structures will be quite heavily used.

File: pkg/core/mpt/node.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/mpt/node.go:24]: Fixed.

File: pkg/core/mpt/branch.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/mpt/branch.go:17]: Fixed.

**Comment by fyrchik** [comment]
Closed in favor of #990.
I have no permission to make `neox-2.x` protected.

**Comment by roman-khimov** [comment]
I'll do that, np.

**Pull request closed without merging**